feat: Support environent variables in config

7rulnik opened the pull request:
**What's the problem this PR addresses?**

Lack of support of environment variables in config files https://github.com/yarnpkg/berry/issues/1146

**How did you fix it?**

Stringify yarnrc configs, split it by variables, replace with value from `process.env`, and fallback if it's not set.

I took fallback algorythm from docker-compose:

>Added support for shell-style inline defaults in variable interpolation.
>The supported forms are ${FOO-default} (fall back if FOO is unset) and
> ${FOO:-default} (fall back if FOO is unset or empty).

If you fine with this implementation I will add tests and will document this. Also I could rewrite it to deep-map and map strings only. It should be a better variant

Note that regexp named groups have limited support: http://kangax.github.io/compat-table/es2016plus/#test-RegExp_named_capture_groups. If this is a problem I can refactor it

arcanis commented:
Hey! Thanks for tackling this. I think we wouldn't want to apply this as a pre-processing pass, otherwise it opens the door to pretty weird things if the environment variables contain things that could lead to valid YAML.

What do you think of instead doing this replacement within `interpretValue`, right here:
https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-core/sources/Configuration.ts#L434

7rulnik commented:
Hmm, didn't see it. Will dig into it.
Btw, is there a way to run yarn with node `--inspect` flag?

arcanis commented:
easiest way is `YARN_IGNORE_PATH=1 node --inspect-brk ./scripts/run-yarn.js`

7rulnik commented:
@arcanis I moved replacement into `interpretValue`, refactored replacement to `.replace` method, added tests and documented this.

Not sure about tests: maybe we should separate it and check all cases in `packages/yarnpkg-parsers/tests/syml.test.js` and in `packages/yarnpkg-core/tests/Configuration.test.ts` just check that replacement works?

paul-soporan commented on packages/gatsby/src/pages/configuration/yarnrc.json:
```diff
@@ -1,7 +1,7 @@
   "title": "JSON Schema for Yarnrc files",
   "$schema": "https://json-schema.org/draft/2019-09/schema#",
-  "description": "Yarnrc files (named this way because they must be called `.yarnrc.yml`) are the one place where you'll be able to configure Yarn's internal settings. While Yarn will automatically find them in the parent directories, they should usually be kept at the root of your project (often your repository). Starting from the v2, they **must** be written in valid Yaml and have the right ext ...
+  "description": "Yarnrc files (named this way because they must be called `.yarnrc.yml`) are the one place where you'll be able to configure Yarn's internal settings. While Yarn will automatically find them in the parent directories, they should usually be kept at the root of your project (often your repository). Starting from the v2, they **must** be written in valid Yaml and have the right ext ...
```
This sounds a bit off to me.

arcanis commented:
Perfect! Thanks @7rulnik! 😃

yinzara commented:
@arcanis Unfortunately this has caused a serious issue with conversion of my current package.json's to Yarn 2.0

I use "cross-env" to support environment variable expansion. Because yarn now automatically expands all environment variables in scripts, this is breaking features that "cross-env" has. For example "$PWD" should be replaced with the current working directory, however now when using yarn 2.0 on windows: "cross-env echo $PWD" throws a `Internal Error: Unbound variable "PWD"` and unfortunately there is no way for me to correct the issue through putting a default value in.

yinzara commented:
Actually there are a whole host of use cases this breaks now.

Because you're using a regular expression to evaluate the environment variables, you cannot properly handle nested values and know where the right curly braces ends.

I think you're going to have to come up with another solution for this.  The current one, though it seems great, is going to cause serious issues with people on different environments.

https://github.com/kentcdodds/cross-env/pull/219

yinzara commented:
Could we have some way to simply disable this functionality all together maybe via .yarnrc.yml or via an environment variable?

arcanis commented:
Please open an issue with a repro. This PR is about the configuration, whereas you're talking about scripts 🙂

paul-soporan commented:
Not sure what `cross-env` has to do with the feature introduced in this PR. This PR adds support for environment variables inside the configuration (`.yarnrc.yml`), so that the following can work:
```yml
npmRegistries:
  "//registry.npmjs.org":
    "npmAuthToken": ${MY_AUTH_TOKEN}
```

I believe you're referring to the environment variables inside our portable shell, which go through a [completely different parser](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-parsers/sources/grammars/shell.pegjs) that correctly handles nested environment variables and a lot of other things. The portable shell has support for many features that work across all environments (including on Windows), like environment variables, command chains, subshells, some builtins.

For your use case, you can:
- `cross-env 'echo $PWD'` - quote what needs to be passed to `cross-env` so that our shell doesn't expand it
- `echo $(pwd)` - use a subshell and the `pwd` builtin which works across all environments

yinzara commented:
THANKS! That's good enough! I appreciate it.

paul-soporan commented:
I guess we can also add support for the `$PWD` environment variable for better compliance with bash, even though we'd introduce 2 ways to do the same thing (`echo $(pwd)` and `echo $PWD`). @arcanis What do you think?

yinzara commented:
I'll open another issue specifically to add $PWD resolution so it's bash compatible.  I think that's an important thing.
